sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java

```diff
@@ -86,12 +89,19 @@ public Node pushNode(String name, PInput input, PTransform<?, ?> transform) {
   public void finishSpecifyingInput() {
     // Inputs must be completely specified before they are consumed by a transform.
     for (TaggedPValue inputValue : current.getInputs()) {
-      inputValue.getValue().finishSpecifying();
+      Node producerNode = getProducer(inputValue.getValue());
+      PInput input = producerInput.remove(inputValue.getValue());
+      inputValue.getValue().finishSpecifying(input, producerNode.getTransform());
+      checkState(
+          producers.get(inputValue.getValue()) != null,
+          "Producer unknown for input %s",
+          inputValue);
       checkState(
           producers.get(inputValue.getValue()) != null,
           "Producer unknown for input %s",
           inputValue.getValue());
     }
+    current.input.finishSpecifying();
```

Comment:
I think you can delete this and the method itself. I went so far as to load it up in IntelliJ and confirm this is the only reference to the method. Have I missed something?